pkg/reconciler/pipelinerun/resources/pipelinerunstate.go

```diff
@@ -193,53 +192,30 @@ func (facts *PipelineRunFacts) GetPipelineConditionStatus(pr *v1beta1.PipelineRu
 		}
 	}
 
-	allTasks := []string{}
-	withStatusTasks := []string{}
-	skipTasks := []v1beta1.SkippedTask{}
-	failedTasks := int(0)
-	cancelledTasks := int(0)
-	reason := v1beta1.PipelineRunReasonSuccessful.String()
-
-	// Check to see if all tasks are success or skipped
-	//
-	// The completion reason is also calculated here, but it will only be used
-	// if all tasks are completed.
-	//
-	// The pipeline run completion reason is set from the taskrun completion reason
-	// according to the following logic:
-	//
-	// - All successful: ReasonSucceeded
-	// - Some successful, some skipped: ReasonCompleted
-	// - Some cancelled, none failed: ReasonCancelled
-	// - At least one failed: ReasonFailed
-	for _, rprt := range facts.State {
-		allTasks = append(allTasks, rprt.PipelineTask.Name)
-		switch {
-		case rprt.IsSuccessful():
-			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
-		case rprt.Skip(facts):
-			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
-			skipTasks = append(skipTasks, v1beta1.SkippedTask{Name: rprt.PipelineTask.Name})
-			// At least one is skipped and no failure yet, mark as completed
-			if reason == v1beta1.PipelineRunReasonSuccessful.String() {
-				reason = v1beta1.PipelineRunReasonCompleted.String()
-			}
-		case rprt.IsCancelled():
-			cancelledTasks++
-			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
-			if reason != v1beta1.PipelineRunReasonFailed.String() {
-				reason = v1beta1.PipelineRunReasonCancelled.String()
-			}
-		case rprt.IsFailure():
-			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
-			failedTasks++
-			reason = v1beta1.PipelineRunReasonFailed.String()
-		}
-	}
+	// report the count in PipelineRun Status
+	//  get the count of successful tasks, failed tasks, cancelled tasks, skipped task, and incomplete tasks
+	sTasks, fTasks, cTasks, skTasks, iTasks := facts.getPipelineTasksCount()
+	// completed task is a collection of successful, failed, cancelled tasks (skipped tasks are reported separately)
+	cmTasks := sTasks + fTasks + cTasks
 
-	if reflect.DeepEqual(allTasks, withStatusTasks) {
+	// The completion reason is set from the TaskRun completion reason
+	// by default, set it to ReasonRunning
+	reason := v1beta1.PipelineRunReasonRunning.String()
+
+	if facts.checkAllTasksDone() {
```

Comment:
can this check be replaced by `iTasks (incomplete tasks) == 0`? 🤔

Reply:
Great suggestion @jerop, it can be 😜 and it simplifies the check further 🙏